File: ActorClient/ActorClient/CocoaMessenger.m
//
//  CocoaConfigurationBuilder.m
//  ActorClient
//
//  Created by Антон Буков on 23.02.15.
//  Copyright (c) 2015 Anton Bukov. All rights reserved.
//

#import <MagicalRecord/CoreData+MagicalRecord.h>
#import <MMMutableMethods/NSObject+MMAnonymousClass.h>
#import "im/actor/model/cocoa/CocoaThreading.h"
#import "im/actor/model/ConfigurationBuilder.h"

#import "CocoaLogger.h"
#import "CocoaLocale.h"
#import "CocoaMainThread.h"
#import "CocoaNetworking.h"
#import "CocoaStorage.h"
#import "CocoaPhoneBookProvider.h"
#import "CocoaCryptoProvider.h"

#import "CocoaMessenger.h"

@interface CocoaMessenger ()

@end

@implementation CocoaMessenger

- (instancetype)init
{
    self = [super initWithConfig:^{
        AMConfigurationBuilder *confBuilder = [[AMConfigurationBuilder alloc] init];
        [confBuilder setLog:[[CocoaLogger alloc] init]];
        [confBuilder setNetworking:[[CocoaNetworking alloc] init]];
        [confBuilder setThreading:[[AMCocoaThreading alloc] init]];
        [confBuilder setStorage:[[CocoaStorage alloc] init]];
        [confBuilder setMainThread:[[CocoaMainThread alloc] init]];
        [confBuilder addEndpoint:@"tls://mtproto-api.actor.im:443"];
        //[confBuilder addEndpoint:@"tcp://mtproto-api.actor.im:8080"];
        [confBuilder setLocale:[[CocoaLocale alloc] init]];
        [confBuilder setPhoneBookProviderWithAMPhoneBookProvider:[[CocoaPhoneBookProvider alloc] init]];
        [confBuilder setCryptoProviderWithAMCryptoProvider:[[CocoaCryptoProvider alloc] init]];
        return [confBuilder build];
    }()];
    return self;
}

+ (instancetype)messenger
{
    static CocoaMessenger *msgr = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        msgr = [[CocoaMessenger alloc] init];
    });
    return msgr;
}

@end

File: ActorClient/ActorClient/CocoaMessenger.swift
//
//  CocoaMessenger.swift
//  ActorClient
//
//  Created by Stepan Korshakov on 10.03.15.
//  Copyright (c) 2015 Anton Bukov. All rights reserved.
//

import Foundation

private var holder:CocoaMessenger?;

var MSG : CocoaMessenger {
get{
    if (holder == nil){
        var dbPath = NSSearchPathForDirectoriesInDomains(.DocumentDirectory,
            .UserDomainMask, true)[0].stringByAppendingPathComponent("actor.db");
//        var db = FMDatabase(path: dbPath);
//        db.open()
        var builder = AMConfigurationBuilder();
    

        // Providers
        
        builder.setLogProvider(CocoaLogProvider())
        builder.setNetworkProvider(SwiftCocoaNetworkProvider())
        builder.setThreadingProvider(AMCocoaThreadingProvider())
        builder.setStorageProvider(CocoaStorage(dbPath: dbPath))
        builder.setMainThreadProvider(CocoaMainThreadProvider())
        builder.setLocaleProvider(CocoaLocale())
        builder.setPhoneBookProvider(CocoaPhoneBookProvider())
        builder.setCryptoProvider(BCBouncyCastleProvider())
        builder.setFileSystemProvider(CocoaFileSystem())
        builder.setNotificationProvider(iOSNotificationProvider())
        builder.setEnableNetworkLogging(true)
        
        // Connection
        builder.addEndpoint("tls://mtproto-api.actor.im:443");
        
        var value: UInt8 = 0xFF
        var convHash = IOSByteArray.newArrayWithLength(32)
        var buf = UnsafeMutablePointer<UInt8>(convHash.buffer());
        for i in 1..<32 {
            buf.memory = UInt8(arc4random_uniform(255));
            buf++;
//            convHash.buffer()[i] = jbyte(0xFF);
        }
        
        builder.setApiConfiguration(AMApiConfiguration(NSString: "Actor iOS", withInt: 1, withNSString: "???", withNSString: "My Device", withByteArray: convHash))

        holder = CocoaMessenger(AMConfiguration: builder.build());
    }
    return holder!;
    }
}

@objc class CocoaMessenger : AMMessenger {
    class func messenger() -> CocoaMessenger { return MSG }
}